text/0000-mem-replace-with.md

```diff
@@ -0,0 +1,64 @@
+- Feature Name: mem-replace-with
+- Start Date: 2016-09-01
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add a new function to `core::mem` (and thus `std::mem`), `replace_with`, which
+invokes a closure that takes temporary ownership over some value behind a
+mutable reference.
+
+This is similar to the functions provided by [the take_mut crate](https://crates.io/crates/take_mut).
+
+# Motivation
+[motivation]: #motivation
+
+A common pattern, especially for low-level programmers, is to acquire ownership
+of a value behind a reference without immediately giving it a replacement value.
+
+`core::mem::replace_with` allows such thing by generalizing
+`core::mem::replace`, to allow the replacement value being generated by a
+closure and even depend on the old value.
+
+Unfortunately, there are no safe implementation in the standard library,
+leading some Rust users to either reimplement a safe interface for such a
+feature (often flawed due to unwinding, see below) or use things like
+`ptr::read`.
+
+This should standardize the ecosystem with respect to that.
+
+# Detailed design
+[design]: #detailed-design
+
+A complete implementation can be found [here](https://github.com/rust-lang/rust/pull/36186). This implementation is pretty simple, with only one trick which is crucial to safety: handling unwinding.
```

Comment:
Should we just include the code here? The implementation is short, and including makes the RFC self-contained.